File: backend-python/app/utils/id_card_validator.py
# -*- coding: utf-8 -*-
"""
身份证号码强认证工具
包含格式验证、校验码验证、出生日期验证、性别验证
"""

from typing import Tuple, Optional
from datetime import datetime


WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
CHECK_CODES = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2']


def validate_id_card(id_card: str) -> Tuple[bool, str, Optional[str], Optional[str]]:
    """
    验证身份证号码
    
    Args:
        id_card: 身份证号码
        
    Returns:
        Tuple[是否有效, 错误信息/成功信息, 出生日期, 性别]
    """
    if not id_card:
        return False, '身份证号码不能为空', None, None
    
    id_card = id_card.upper().strip()
    
    if len(id_card) != 18:
        return False, '身份证号码必须为18位', None, None
    
    import re
    if not re.match(r'^\d{17}[\dX]$', id_card):
        return False, '身份证号码格式不正确，前17位必须为数字，最后一位可以是数字或X', None, None
    
    total = 0
    for i in range(17):
        total += int(id_card[i]) * WEIGHTS[i]
    check_code = CHECK_CODES[total % 11]
    
    if id_card[17] != check_code:
        return False, '身份证号码校验码错误，请检查是否输入正确', None, None
    
    try:
        year = int(id_card[6:10])
        month = int(id_card[10:12])
        day = int(id_card[12:14])
    except ValueError:
        return False, '身份证号码中日期格式错误', None, None
    
    birth_date = f"{year}-{month:02d}-{day:02d}"
    
    if month < 1 or month > 12:
        return False, '身份证号码中月份无效，应为01-12', None, None
    
    import calendar
    _, days_in_month = calendar.monthrange(year, month)
    if day < 1 or day > days_in_month:
        return False, f'身份证号码中日期无效，{year}年{month}月没有{day}日', None, None
    
    try:
        birth_date_obj = datetime(year, month, day)
        today = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
        
        if birth_date_obj > today:
            return False, '身份证号码中出生日期不能晚于今天', None, None
    except ValueError as e:
        return False, f'身份证号码中出生日期无效: {str(e)}', None, None
    
    min_year = 1900
    if year < min_year:
        return False, f'身份证号码中出生年份不能早于{min_year}年', None, None
    
    gender_code = int(id_card[16:17])
    gender = '男' if gender_code % 2 == 1 else '女'
    
    return True, '身份证号码验证通过', birth_date, gender


def extract_birth_date(id_card: str) -> Optional[str]:
    """
    从身份证号码提取出生日期
    
    Args:
        id_card: 身份证号码
        
    Returns:
        出生日期字符串 (YYYY-MM-DD格式) 或 None
    """
    if not id_card or len(id_card) != 18:
        return None
    
    year = id_card[6:10]
    month = id_card[10:12]
    day = id_card[12:14]
    
    return f"{year}-{month}-{day}"


def extract_gender(id_card: str) -> Optional[str]:
    """
    从身份证号码提取性别
    
    Args:
        id_card: 身份证号码
        
    Returns:
        性别 ('男' 或 '女') 或 None
    """
    if not id_card or len(id_card) != 18:
        return None
    
    gender_code = int(id_card[16:17])
    return '男' if gender_code % 2 == 1 else '女'


def calculate_age(birth_date: str) -> Optional[int]:
    """
    计算年龄
    
    Args:
        birth_date: 出生日期字符串 (YYYY-MM-DD格式)
        
    Returns:
        年龄 或 None
    """
    if not birth_date:
        return None
    
    try:
        birth = datetime.strptime(birth_date, '%Y-%m-%d')
        today = datetime.now()
        
        age = today.year - birth.year
        if (today.month, today.day) < (birth.month, birth.day):
            age -= 1
        
        return age if age >= 0 else None
    except ValueError:
        return None

File: H5/src/utils/idCardValidator.ts
/**
 * 身份证号码强认证工具
 * 包含格式验证、校验码验证、出生日期验证、性别验证
 */

// TODO: 身份证验证 - 考虑加入有效期验证
// FIXME: 目前只支持18位身份证，15位的需要兼容
// TODO: 考虑加入港澳台身份证验证

export interface IdCardValidationResult {
  valid: boolean
  message: string
  birthDate?: string
  gender?: string
}

const WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
const CHECK_CODES = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2']

/**
 * 验证身份证号码
 * @param idCard 身份证号码
 * @returns 验证结果，包含是否有效、错误信息、出生日期、性别
 */
export function validateIdCard(idCard: string): IdCardValidationResult {
  if (!idCard) {
    return { valid: false, message: '身份证号码不能为空' }
  }

  const idCardUpper = idCard.toUpperCase().trim()

  if (idCardUpper.length !== 18) {
    return { valid: false, message: '身份证号码必须为18位' }
  }

  if (!/^\d{17}[\dX]$/.test(idCardUpper)) {
    return { valid: false, message: '身份证号码格式不正确，前17位必须为数字，最后一位可以是数字或X' }
  }

  let sum = 0
  for (let i = 0; i < 17; i++) {
    sum += parseInt(idCardUpper[i]!, 10) * WEIGHTS[i]!
  }
  const checkCode = CHECK_CODES[sum % 11]
  
  if (idCardUpper[17] !== checkCode) {
    return { valid: false, message: '身份证号码校验码错误，请检查是否输入正确' }
  }

  const year = parseInt(idCardUpper.substring(6, 10), 10)
  const month = parseInt(idCardUpper.substring(10, 12), 10)
  const day = parseInt(idCardUpper.substring(12, 14), 10)

  const birthDate = `${year}-${month.toString().padStart(2, '0')}-${day.toString().padStart(2, '0')}`

  if (month < 1 || month > 12) {
    return { valid: false, message: '身份证号码中月份无效，应为01-12' }
  }

  const daysInMonth = new Date(year, month, 0).getDate()
  if (day < 1 || day > daysInMonth) {
    return { valid: false, message: `身份证号码中日期无效，${year}年${month}月没有${day}日` }
  }

  const birthDateObj = new Date(year, month - 1, day)
  const today = new Date()
  today.setHours(0, 0, 0, 0)
  
  if (birthDateObj > today) {
    return { valid: false, message: '身份证号码中出生日期不能晚于今天' }
  }

  const minYear = 1900
  if (year < minYear) {
    return { valid: false, message: `身份证号码中出生年份不能早于${minYear}年` }
  }

  const genderCode = parseInt(idCardUpper.substring(16, 17), 10)
  const gender = genderCode % 2 === 1 ? '男' : '女'

  return {
    valid: true,
    message: '身份证号码验证通过',
    birthDate,
    gender
  }
}

/**
 * 从身份证号码提取出生日期
 * @param idCard 身份证号码
 * @returns 出生日期字符串 (YYYY-MM-DD格式)
 */
export function extractBirthDate(idCard: string): string | null {
  if (!idCard || idCard.length !== 18) {
    return null
  }
  
  const year = idCard.substring(6, 10)
  const month = idCard.substring(10, 12)
  const day = idCard.substring(12, 14)
  
  return `${year}-${month}-${day}`
}

/**
 * 从身份证号码提取性别
 * @param idCard 身份证号码
 * @returns 性别 ('男' 或 '女')
 */
export function extractGender(idCard: string): string | null {
  if (!idCard || idCard.length !== 18) {
    return null
  }
  
  const genderCode = parseInt(idCard.substring(16, 17), 10)
  return genderCode % 2 === 1 ? '男' : '女'
}

/**
 * 计算年龄
 * @param birthDate 出生日期字符串 (YYYY-MM-DD格式)
 * @returns 年龄
 */
export function calculateAge(birthDate: string): number | null {
  if (!birthDate) {
    return null
  }
  
  const birth = new Date(birthDate)
  const today = new Date()
  
  let age = today.getFullYear() - birth.getFullYear()
  const monthDiff = today.getMonth() - birth.getMonth()
  
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
    age--
  }
  
  return age >= 0 ? age : null
}

/**
 * 获取身份证号码所属地区码（前6位）
 * @param idCard 身份证号码
 * @returns 地区码
 */
export function getRegionCode(idCard: string): string | null {
  if (!idCard || idCard.length < 6) {
    return null
  }
  return idCard.substring(0, 6)
}
